pkg/scheduler/algorithm/predicates/predicates.go

```diff
@@ -1704,3 +1706,68 @@ func (c *VolumeBindingChecker) predicate(pod *v1.Pod, meta PredicateMetadata, no
 	klog.V(5).Infof("All PVCs found matches for pod %v/%v, node %q", pod.Namespace, pod.Name, node.Name)
 	return true, nil, nil
 }
+
+// EvenPodsSpreadPredicate checks if a pod can be scheduled on a node which satisfies
+// its topologySpreadConstraints.
+func EvenPodsSpreadPredicate(pod *v1.Pod, meta PredicateMetadata, nodeInfo *schedulernodeinfo.NodeInfo) (bool, []PredicateFailureReason, error) {
+	node := nodeInfo.Node()
+	if node == nil {
+		return false, nil, fmt.Errorf("node not found")
+	}
+
+	constraints := getHardTopologySpreadConstraints(pod)
+	if len(constraints) == 0 {
+		return true, nil, nil
+	}
+
+	var topologyPairsPodSpreadMap *topologyPairsPodSpreadMap
+	if predicateMeta, ok := meta.(*predicateMetadata); ok {
+		topologyPairsPodSpreadMap = predicateMeta.topologyPairsPodSpreadMap
+	} else { // We don't have precomputed metadata. We have to follow a slow path to check spread constraints.
+		// TODO(Huang-Wei): get it implemented
+		return false, nil, errors.New("metadata not pre-computed for EvenPodsSpreadPredicate")
+	}
+
+	if topologyPairsPodSpreadMap == nil || len(topologyPairsPodSpreadMap.minMatchMap) == 0 {
+		return true, nil, nil
+	}
+
+	selfMatch, err := podLabelsMatchesSpreadConstraints(pod.Labels, constraints)
+	if err != nil {
+		return false, nil, err
+	}
+	selfMatchNum := 0
+	if selfMatch {
+		selfMatchNum = 1
+	}
+	for _, constraint := range constraints {
+		tpKey := constraint.TopologyKey
+		tpVal, ok := node.Labels[constraint.TopologyKey]
+		if !ok {
+			klog.V(5).Infof("node '%s' doesn't have required label '%s'", node.Name, tpKey)
+			return false, []PredicateFailureReason{ErrTopologySpreadConstraintsNotMatch}, nil
+		}
+
+		pair := topologyPair{key: tpKey, value: tpVal}
+		minMatchNum, ok := topologyPairsPodSpreadMap.minMatchMap[tpKey]
+		if !ok {
+			// error which should not happen
+			klog.Errorf("internal error: get minMatchNum from key %q of %#v", tpKey, topologyPairsPodSpreadMap.minMatchMap)
+			continue
+		}
+		// judging criteria:
+		// 'existing matching num' + 'if self-match (1 or 0)' - 'global min matching num' <= 'maxSkew'
+		matchNum := len(topologyPairsPodSpreadMap.topologyPairToPods[pair])
+
+		// TODO(Huang-Wei): remove this after thorough testing
+		// fmt.Printf("node '%s' => spreadConstraint[%s]: matchNum(%d) + selfMatchNum(%d) - minMatchNum(%d) ?<= maxSkew(%d)\n", node.Name, tpKey, matchNum, selfMatchNum, minMatchNum, constraint.MaxSkew)
+
+		// TODO(Huang-Wei): check if it can overflow?
+		if int32(matchNum+selfMatchNum)-minMatchNum > constraint.MaxSkew {
```

Comment:
Can you please add an example like the one you have in the KEP to document the reasoning behind this logic.

I would also compute the skew on a separate line above. 

```
skew := matchNum+selfMatchNum-minMatchNum
if skew > constraint.MaxSkew {
```

btw, is the casting to int32 actually necessary?

Reply:
Sure, will do. (I will start to address comments of this PR after the previous PR gets stable. Otherwise, I will fall into endless "rebasing" labor work..)

> btw, is the casting to int32 actually necessary?

Yes. Golang complains if int32 and int are combined used: `invalid operation blabla: mismatched types int32 and int`.